### ISSUE TITLE
k_v in decoder have shared projection?

### ISSUE THREAD

**Issue by fawazsammani** [husthuaan/AoANet#22]
Hi @husthuaan .
I'm going through your code. Thanks for uploading it! In section 3.2.2 in your paper, you show in the equation that there is a separate projection (separate weights) for the keys and values (which are your image features) in the **decoder part.** However, in your code, you set `project_k_v=0 `in the decoder and project both of them using:
` self.ctx2att = nn.Linear(opt.rnn_size, 2 * opt.multi_head_scale * opt.rnn_size)`
So both the K and V have a shared projection right?

**Comment by husthuaan** [comment]
self.ctx2att outputs a vector of size 2*rnn_size, where one half is for K and the other half is for V. Thus they are still different projections.

**Comment by fawazsammani** [comment]
Thanks for your reply. I didnt inspect that the line `p_att_feats.narrow(2, 0, self.multi_head_scale * self.d_model)` splits them.

**Comment by fawazsammani** [comment]
@husthuaan I also want to ask. In the paper, you state tha you used the Adam Optimizer and annealed the learning rate: 

> ADAM [23] optimizer is used with a
> learning rate initialized by 2e-4 and annealed by 0.8 every 3
> epochs

In the code, I cannot see any option for `noamopt` in `train.sh`. But you do specify that you use noamopt as you've added it to the list of models. May I know if you've used noamopt (with no learning rate annealing) or Adam (with learning rate annealing)?

**Comment by husthuaan** [comment]
The training script uses pure Adam with learning rate annealing, just as the paper states.
Using noamopt or not is an option, and you can specify it in `train.sh`.

**Comment by fawazsammani** [comment]
Thanks!
